### ISSUE TITLE
choice of network output and NeuralNet criterion

### ISSUE THREAD

**Issue by FlorinAndrei** [skorch-dev/skorch#830]
I use PyTorch for classification with one-hot-encoded labels. I'm looking at this page:

https://skorch.readthedocs.io/en/stable/user/neuralnet.html#criterion

It says:

> When you use the NeuralNetClassifier, the criterion is set to PyTorch NLLLoss by default. Furthermore, if you don’t change it loss to another criterion, NeuralNetClassifier assumes that the module returns probabilities and will automatically apply a logarithm on them (which is what NLLLoss expects).

The quickstart example in the docs suggests using F.softmax() on the output of the classifier network, and uses the default value for the criterion in NeuralNetClassifier() - which seems to be NLLLoss().

When training PyTorch with manually-selected hyperparameters, the combinations that are most often used are:

- linear output + nn.CrossEntropyLoss
- log_softmax output + nn.NLLLoss

What if I want to write a model just once, and then pass it without changes to Skorch for hyperparameter tuning, and then again train it on a data set, without any extra transformations of the data, the best option seem to be to just use a linear output with CrossEntropyLoss. If I have softmax on the output, then in regular training I'd have to log the output, then pass it to NLLLoss(), which seems like an unnecessary extra step.

Why not just recommend using a linear output by default, and have the default criterion in Skorch be CrossEntropyLoss?

**Comment by BenjaminBossan** [comment]
> When training PyTorch with manually-selected hyperparameters, the combinations that are most often used are:
> 
>     * linear output + nn.CrossEntropyLoss
> 
>     * log_softmax output + nn.NLLLoss

> Why not just recommend using a linear output by default, and have the default criterion in Skorch be CrossEntropyLoss?

These are some valid points. We have the defaults as they currently are for historical reasons. When skorch was released (July 2017), there were not really any established pytorch standards, but there were standards from other libraries like nolearn and keras. With hindsight, we might have chosen the standard you suggested.

At this point, changing such a fundamental default might not be worth it, it's hard for us to know how much code would be affected. And it's really not hard to make the change you suggested with an extra line of code. That's why we decided to keep it as is.

**Comment by FlorinAndrei** [comment]
I see. In that case, could you add some notes to the Skorch documentation - basically saying "here are the current best practices" etc.

The Quickstart page recommends a softmax() output, but that's not the best choice given the default criterion used by Skorch.

https://skorch.readthedocs.io/en/stable/user/quickstart.html

Maybe, on that page, suggest one of the standard combos (log_softmax and NLLLoss) in the code, and mention the other combo (linear and CrossEntropyLoss) in the text?

**Comment by BenjaminBossan** [comment]
Yes, we could add something to the docs to that effect.

Just so that we're on the same page: When `NLLLoss` is used, skorch takes the log of the softmax output for the user. The only problem with that, as far as I'm aware, is that softmax + log is not as numerically as using logsoftmax directly. Are there other reasons to be aware of?

**Comment by FlorinAndrei** [comment]
I'm not aware of any, but I'm mostly interested in just having a linear output and using crossentropy with that.

> Yes, we could add something to the docs to that effect.

Thanks!

**Comment by BenjaminBossan** [comment]
Solved via #852
